src/NUnitFramework/framework/Internal/Results/TestSuiteResult.cs

```diff
@@ -112,46 +117,53 @@ public override IList<ITestResult> Children
         /// <param name="result">The result to be added</param>
         public virtual void AddResult(ITestResult result)
         {
-            Children.Add(result);
+            Children.Enqueue(result);
 
             //AssertCount += result.AssertCount;
 
-            // If this result is marked cancelled, don't change it
-            if (ResultState != ResultState.Cancelled)
+            bool stateSet;
+            do
+            {
+                stateSet = true;
+
+                var resultState = ResultState;
+
+                // If this result is marked cancelled, don't change it
+                if (resultState == ResultState.Cancelled)
+                    break;
+
                 switch (result.ResultState.Status)
                 {
                     case TestStatus.Passed:
 
-                        if (ResultState.Status == TestStatus.Inconclusive)
-                            SetResult(ResultState.Success);
+                        if (resultState.Status == TestStatus.Inconclusive)
+                            stateSet = SetResultIf(resultState, ResultState.Success);
 
                         break;
 
                     case TestStatus.Failed:
 
 
-                        if (ResultState.Status != TestStatus.Failed)
-                            SetResult(ResultState.ChildFailure, CHILD_ERRORS_MESSAGE);
+                        if (resultState.Status != TestStatus.Failed)
+                            stateSet = SetResultIf(resultState, ResultState.ChildFailure, CHILD_ERRORS_MESSAGE);
 
                         break;
 
                     case TestStatus.Skipped:
 
                         if (result.ResultState.Label == "Ignored")
-                            if (ResultState.Status == TestStatus.Inconclusive || ResultState.Status == TestStatus.Passed)
-                                SetResult(ResultState.Ignored, CHILD_IGNORE_MESSAGE);
-
-                        break;
+                            if (resultState.Status == TestStatus.Inconclusive || resultState.Status == TestStatus.Passed)
+                                stateSet = SetResultIf(resultState, ResultState.Ignored, CHILD_IGNORE_MESSAGE);
 
-                    default:
                         break;
                 }
+            } while (!stateSet);
 
-            AssertCount += result.AssertCount;
-            _passCount += result.PassCount;
-            _failCount += result.FailCount;
-            _skipCount += result.SkipCount;
-            _inconclusiveCount += result.InconclusiveCount;
+            Interlocked.Add (ref InternalAssertCount, result.AssertCount);
+            Interlocked.Add (ref _passCount, result.PassCount);
+            Interlocked.Add (ref _failCount, result.FailCount);
+            Interlocked.Add (ref _skipCount, result.SkipCount);
+            Interlocked.Add (ref _inconclusiveCount, result.InconclusiveCount);
         }
 
```

Comment:
Looking it over, I can see that there is currently no race condition. AddResult performs the following seven operations (each) atomically:
1. Actually adds the child result to the collection.
2. Sets the overall result, based on the child result
3-7. Increments each of the counters separately.

All these things happen independently and the test in SelectIf ensures that the code will be retried if another thread performs step 1 before step 2 can be finished.

However, the state of the result is invalid between the steps. Between 1 and 2, the overall resultstate is incorrect. After 2 but before 7, the totals don't match. If anyone accesses the result during that time, it's state wil be invalid. Admittedly, this is exactly the case with the original code.

The key problems I see are

1. The code is (as you say) complex and may be hard for some contributors to follow. A simple lock would be easier. If we keep it as it is, I would want us to take great care in reviewing any changes.

2. The performance advantage seems questionable to me given how little AddResult actually does and the limitations of how and when it is called. I assume that the ConcurrentQueue is locking already.

3. When I talk about a lock, I don't necessarily mean the obvious lock/monitor implementation. This is all local to a process and AppDomain and could be optimized in a few ways.

4. The steps involved in AddResult are never supposed to happen independently, so there is no particular need for them to be separately thread-safe. The overall design is based on the idea that results should be added atomically.